apps/wallet/src/dapp-interface/WalletStandardInterface.ts

```diff
@@ -0,0 +1,185 @@
+
+import icon from '../manifest/icons/sui-icon-128.png';
+import { mapToPromise } from './utils';
+import { createMessage } from '_messages';
+import { WindowMessageStream } from '_messaging/WindowMessageStream';
+import { type Payload } from '_payloads';
+import {
+    type AcquirePermissionsRequest,
+    type AcquirePermissionsResponse,
+    ALL_PERMISSION_TYPES,
+} from '_payloads/permissions';
+
+import type { GetAccount } from '_payloads/account/GetAccount';
+import type { GetAccountResponse } from '_payloads/account/GetAccountResponse';
+import type {
+    ExecuteTransactionRequest,
+    ExecuteTransactionResponse,
+} from '_payloads/transactions';
+
+type WalletEventsMap = {
+    [E in WalletEventNames]: Parameters<WalletEvents[E]>[0];
+};
+
+// TODO: rebuild event interface with Mitt.
+export class SuiWallet implements Wallet {
+    readonly #events: Emitter<WalletEventsMap>;
+    readonly #version = '1.0.0' as const;
+    readonly #name = 'Sui Wallet' as const;
+    #account: ReadonlyWalletAccount | null;
+    #messagesStream: WindowMessageStream;
+
+    get version() {
+        return this.#version;
+    }
+
+    get name() {
+        return this.#name;
+    }
+
+    get icon() {
+        // TODO: Improve this with ideally a vector logo.
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        return icon as any;
+    }
+
+    get chains() {
+        // TODO: Extract chain from wallet:
+        return SUI_CHAINS;
+    }
+
+    get features(): ConnectFeature & SuiSignAndExecuteTransactionFeature {
+        return {
+            'standard:connect': {
+                version: '1.0.0',
+                connect: this.#connect,
+            },
+            'sui:signAndExecuteTransaction': {
+                version: '1.0.0',
+                signAndExecuteTransaction: this.#signAndExecuteTransaction,
+            },
+        };
+    }
+
+    get accounts() {
+        return this.#account ? [this.#account] : [];
+    }
+
+    constructor() {
+        this.#events = mitt();
+        this.#account = null;
+        this.#messagesStream = new WindowMessageStream(
+            'sui_in-page',
+            'sui_content-script'
+        );
+
+        this.#connected();
+    }
+
+    on<E extends WalletEventNames>(
+        event: E,
+        listener: WalletEvents[E]
+    ): () => void {
+        this.#events.on('standard:change', listener);
+        return () => this.#events.off(event, listener);
+    }
+
+    #connected = async () => {
+        const accounts = await mapToPromise(
+            this.#send<GetAccount, GetAccountResponse>({
+                type: 'get-account',
+            }),
+            (response) => response.accounts
+        );
+
+        const properties: ('accounts' | 'chains')[] = [];
+
+        const [address] = accounts;
+
+        if (address) {
+            const account = this.#account;
+            if (!account || account.address !== address) {
```

Comment:
Bugs hide here for the trigger condition of the accounts change event.
Consider the scenario: The outside dapp firstly connects to Sui wallet and then switch to another wallet. Since there is no such `disconnect` cleanup logic, the internal #account of Sui wallet is still the previous one. And dapp switches to Sui wallet again, the `!account || account.address !== address` would skip the below `standard:change' notification, which is needed by dapp this time.

Reply:
I don’t think the dapp should need to add a listener for the change event, in the case you described the dapp could just inspect the accounts on the wallet and determine it can access accounts.

Reply:
> I don’t think the dapp should need to add a listener for the change event, in the case you described the dapp could just inspect the accounts on the wallet and determine it can access accounts.

You mean dapp inspects the accounts, okay, this could work.